pkg/resource/resource.go

```diff
@@ -243,45 +243,33 @@ func IsConditionTrue(c v1alpha1.Condition) bool {
 
 // An Applicator applies changes to an object.
 type Applicator interface {
-	Apply(context.Context, client.Client, runtime.Object, ...ApplyOption) error
+	Apply(context.Context, client.Client, runtime.Object, ...ApplyOptFn) error
 }
 
 // An ApplyFn is a function that satisfies the Applicator interface.
-type ApplyFn func(context.Context, client.Client, runtime.Object, ...ApplyOption) error
+type ApplyFn func(context.Context, client.Client, runtime.Object, ...ApplyOptFn) error
 
 // Apply changes to the supplied object.
-func (fn ApplyFn) Apply(ctx context.Context, c client.Client, o runtime.Object, ao ...ApplyOption) error {
+func (fn ApplyFn) Apply(ctx context.Context, c client.Client, o runtime.Object, ao ...ApplyOptFn) error {
 	return fn(ctx, c, o, ao...)
 }
 
-// ApplyOptions configure how changes are applied to an object.
-type ApplyOptions struct {
-	// ControllersMustMatch requires any existing object to have a controller
-	// reference, and for that controller reference to match the controller
-	// reference of the supplied object.
-	ControllersMustMatch bool
-}
-
-// An ApplyOption configures how changes are applied to an object.
-type ApplyOption func(a *ApplyOptions)
+// An ApplyOptFn is a function that checks for a condition before patch.
+type ApplyOptFn func(ctx context.Context, current, desired runtime.Object) error
 
 // ControllersMustMatch requires any existing object to have a controller
 // reference, and for that controller reference to match the controller
 // reference of the supplied object.
-func ControllersMustMatch() ApplyOption {
-	return func(a *ApplyOptions) {
-		a.ControllersMustMatch = true
+func ControllersMustMatch(ctx context.Context, current, desired runtime.Object) error {
+	if !meta.HaveSameController(current.(metav1.Object), desired.(metav1.Object)) {
+		return errors.New("existing object has a different (or no) controller")
 	}
+	return nil
 }
```

Comment:
While this pattern works for `ControllersMustMatch` I think we should prefer to have our options _return_ a function with signature `func(ctx context.Context, current, desired runtime.Object) error` rather than _being_ a function with that signature. This would allow us to be consistent if we instead wanted an option like:

```go
func ControllerMustMatch(c corev1.ObjectReference) ApplyOption {
    return func(_ context.Context, current, desired runtime.Object) error {
       // stuff
       return nil
    }
}
```
```